src/bundle/Security/Http/Authenticator/TwoFactorAuthenticator.php

```diff
@@ -0,0 +1,183 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\Http\Authenticator;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorTokenInterface;
+use Scheb\TwoFactorBundle\Security\Authorization\TwoFactorAccessDecider;
+use Scheb\TwoFactorBundle\Security\Http\Authentication\AuthenticationRequiredHandlerInterface;
+use Scheb\TwoFactorBundle\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Scheb\TwoFactorBundle\Security\Http\Authenticator\Passport\Badge\TrustedDeviceBadge;
+use Scheb\TwoFactorBundle\Security\Http\Authenticator\Passport\Credentials\TwoFactorCodeCredentials;
+use Scheb\TwoFactorBundle\Security\Http\Authenticator\Passport\TwoFactorPassport;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorAuthenticationEvent;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorAuthenticationEvents;
+use Scheb\TwoFactorBundle\Security\TwoFactor\TwoFactorFirewallConfig;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationFailureHandlerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+use Symfony\Component\Security\Http\Authenticator\AuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\InteractiveAuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+class TwoFactorAuthenticator implements AuthenticatorInterface, InteractiveAuthenticatorInterface
+{
+    /**
+     * @var TwoFactorFirewallConfig
+     */
+    private $twoFactorFirewallConfig;
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var AuthenticationSuccessHandlerInterface
+     */
+    private $successHandler;
+
+    /**
+     * @var AuthenticationFailureHandlerInterface
+     */
+    private $failureHandler;
+
+    /**
+     * @var AuthenticationRequiredHandlerInterface
+     */
+    private $authenticationRequiredHandler;
+
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+    /**
+     * @var TwoFactorAccessDecider
+     */
+    private $twoFactorAccessDecider;
```

Comment:
Hello @scheb,

I see you inject a `TwoFactorAccessDecider` object, but it is not used in this class.

Reply:
Oh, that must have slipped through in the many refactorings. Thanks for noting that, have removed that unused dependency.